src/AnalysisDD4hep.cxx

```diff
@@ -0,0 +1,735 @@
+		{
+		  double deta = abs(kin->vecHadron.Eta() - mcpart[imc].vecPart.Eta());
+		  if( deta < mineta )
+		    {
+		      mineta = deta;
+		      kinTrue->vecHadron = mcpart[imc].vecPart;
+		    }
+		}
+	    }
+
+	  kinTrue->CalculateHadronKinematics();
+
+	  Double_t w = weight->GetWeight(*kin);
+	  wTotal += w;
+
+	  hpx += reco_px;
+	  hpy += reco_py;
+	  v_had += v_temp;
+
+	  // apply cuts
+	  if(kin->CutFull()) {
+	    // decide which histogram sets to fill
+	    // -- `v_A` will be the list of bins that this event's variable `A`
+	    //    will be a part of; we use these lists to determine the list
+	    //    of histogram sets to fill
+	    // - check pT bin
+	    CheckBins( AN->BinScheme("pt"), v_pt, kin->pT );
+	    CheckBins( AN->BinScheme("x"),  v_x,  kin->x );
+	    CheckBins( AN->BinScheme("z"),  v_z,  kin->z );
+	    CheckBins( AN->BinScheme("q2"), v_q,  kin->Q2 );
+	    CheckBins( AN->BinScheme("y"),  v_y,  kin->y );
+
+	    // build list of histogram sets to fill
+	    histSetFillList.clear();
+	    for(int bpt : v_pt) {
+	      for(int bx : v_x) {
+		for(int bz : v_z) {
+		  for(int bq : v_q) {
+		    for(int by : v_y) {
+		      if(!AN->CheckDiagonal(bpt,bx,bz,bq)) {
+			histSetFillList.push_back(histSet[bpt][bx][bz][bq][by][bFinalState]);
+		      };
+		    };
+		  };
+		};
+	      };
+	    };
+
+	    // loop through list of histogram sets, and fill them
+	    for(Histos *H : histSetFillList) {
+	      // DIS kinematics
+	      dynamic_cast<TH2*>(H->Hist("Q2vsX"))->Fill(kin->x,kin->Q2,w);
+	      H->Hist("Q")->Fill(TMath::Sqrt(kin->Q2),w);
+	      H->Hist("x")->Fill(kin->x,w);
+	      H->Hist("W")->Fill(kin->W,w);
+	      H->Hist("y")->Fill(kin->y,w);
+	      // hadron 4-momentum
+
+	      H->Hist("pLab")->Fill(kin->pLab,w);
+	      H->Hist("pTlab")->Fill(kin->pTlab,w);
+	      H->Hist("etaLab")->Fill(kin->etaLab,w);
+	      H->Hist("phiLab")->Fill(kin->phiLab,w);
+	      // hadron kinematics
+	      H->Hist("z")->Fill(kin->z,w);
+	      H->Hist("pT")->Fill(kin->pT,w);
+	      H->Hist("qT")->Fill(kin->qT,w);
+	      H->Hist("qTq")->Fill(kin->qT/TMath::Sqrt(kin->Q2),w);
+	      H->Hist("mX")->Fill(kin->mX,w);
+	      H->Hist("phiH")->Fill(kin->phiH,w);
+	      H->Hist("phiS")->Fill(kin->phiS,w);
+	      // cross sections (divide by lumi after all events processed)
+	      H->Hist("Q_xsec")->Fill(TMath::Sqrt(kin->Q2),w);
+	      // DIS kinematics resolution
+	      H->Hist("xRes")->Fill(kin->x - kinTrue->x,w);
+	      H->Hist("yRes")->Fill(kin->y - kinTrue->y,w);
+	      H->Hist("Q2Res")->Fill(kin->Q2 - kinTrue->Q2,w);
+	    };
+
+
+	    // fill simple tree (not binned)
+	    // TODO: consider adding a `finalState` cut
+	    if( AN->writeSimpleTree && histSetFillList.size()>0 ) ST->FillTree(w);
+	
+	  }//if cut
+	}//trk loop
+
+      //Hadronic reconstruction method
+      kin->sigmah = (kin->vecElectron.E() - kin->vecElectron.Pz());
+      kin->Pxh = hpx - kin->vecElectron.Px();
+      kin->Pyh = hpy - kin->vecElectron.Py();
```

Comment:
Shouldn't these be set before calling one of the `Kinematics::CalculateDIS` methods? Looking closer in `Kinematics` it seems that `sigmah`, `Pxh`, and `Pyh` never get initialized! https://github.com/c-dilks/largex-eic/issues/22 is opened to take care of that

Reply:
Moved final state loop before calling the DIS kinematics calculation. Fixed by the commit 23d4481.